cli/types.ts

```diff
@@ -0,0 +1,48 @@
+// Copyright 2018-2024 the Deno authors. All rights reserved. MIT license.
+
+type OptionTypeConstructor<T> = T extends boolean ? BooleanConstructor
+  : T extends number ? NumberConstructor
+  : T extends string ? StringConstructor
+  : BooleanConstructor | NumberConstructor | StringConstructor;
+
+export interface Option<T> {
+  name: string;
+  alias?: string;
+  description?: string;
+  default?: T;
+  type?: OptionTypeConstructor<T>;
+  negatable?: boolean;
+  value?: {
+    name: string;
+    optional?: boolean;
+    multiple?: boolean;
+    requireEquals?: boolean;
+  };
+  fn?: (value: T) => T | void;
+}
+
+export interface Argument<T = unknown> {
+  name: string;
+  description?: string;
+  multiple?: boolean;
+  optional?: boolean;
+  fn?: (value: T) => T | void;
+}
+
+export interface Command<T> {
+  name: string;
+  description?: string;
+  options?: ReadonlyArray<Option<T>>;
+  commands?: ReadonlyArray<Command<T>>;
+  arguments?: ReadonlyArray<Argument>;
+  fn?: (result: ParseResult) => void;
+}
```

Comment:
Maybe consider `options` and `commands` to be a `Record` where keys are names?

Reply:
I have considered that, but if an option doesn't have properties, They would need to be declared with an empty object, which seems strange:
```ts
parse(["--foo"], { options: { foo: {} } })
```
```ts
parse(["run", "--foo"], {
  commands: {
    run: { options: { foo: {} } }
  }
});
```
Or is there a better way?